jerry-ext/module/module.c

```diff
@@ -0,0 +1,215 @@
+ * @p module is released.
+ */
+static jerry_value_t
+jerryx_module_add_to_cache (jerry_value_t cache, /**< cache to which to add the module */
+                            jerry_value_t module_name, /**< key at which to cache the module */
+                            jerry_value_t module) /**< the module to cache */
+{
+  jerry_value_t ret = jerry_set_property (cache, module_name, module);
+
+  if (jerry_value_has_error_flag (ret))
+  {
+    jerry_release_value (module);
+  }
+  else
+  {
+    jerry_release_value (ret);
+    ret = module;
+  }
+
+  return ret;
+} /* jerryx_module_add_to_cache */
+
+/**
+ * Declare and define the default module resolver - one which examines what modules are defined in the above linker
+ * section and loads one that matches the requested name, caching the result for subsequent requests using the context
+ * data mechanism.
+ */
+JERRYX_MODULE_RESOLVER (native_resolver)
+{
+  int index;
+  const jerryx_module_t *module_p = NULL;
+  jerry_value_t ret = 0;
+
+  /* Look for the module by its name in the list of module definitions. */
+  for JERRYX_SECTION_ITERATE (jxm_modules, index, module_p)
+  {
+    if (!strcmp ((char *) module_p->name, (char *) name))
+    {
+      /* If we find the module by its name we load it and cache it if it has an on_resolve () and complain otherwise. */
+      ret = (module_p->on_resolve) ? module_p->on_resolve () :
+      jerryx_module_create_error (JERRY_ERROR_TYPE, on_resolve_absent, name);
+      break;
+    }
+  }
+
+  return ret;
+} /* JERRYX_MODULE_RESOLVER */
+
+/**
+ * Declare the section where we expect to find module resolvers - at least the one above will be there.
+ */
+typedef jerry_value_t (*jerryx_module_resolver_t)(const jerry_char_t *);
+JERRYX_SECTION_DECLARE (jxm_resolvers, jerryx_module_resolver_t)
+
+/**
+ * Resolve a single module using the module resolvers available in the section declared above and load it into the
+ * current context.
+ *
+ * @return a jerry_value_t containing one of the followings:
+ *   - the result of having loaded the module named @p name, or
+ *   - the result of a previous successful load, or
+ *   - an error indicating that something went wrong during the attempt to load the module.
+ */
+jerry_value_t
+jerryx_module_resolve (const jerry_char_t *name) /**< name of the module to load */
+{
+  int index;
+  const jerryx_module_resolver_t *resolver_p;
+  jerry_value_t ret;
+  jerry_value_t instances = *(jerry_value_t *) jerry_get_context_data (&jerryx_module_manager);
+  jerry_value_t module_name = jerry_create_string_from_utf8 (name);
+
+  /* Return the cached instance if present. */
+  ret = jerryx_module_check_cache (instances, module_name);
+  if (ret != 0)
+  {
+    goto done;
+  }
+
+  /* Try each resolver in the section until one manages to return non-zero. */
+  for JERRYX_SECTION_ITERATE (jxm_resolvers, index, resolver_p)
```

Comment:
That's true, but, aside from support for ES6 modules syntax (like `import` and `export`) I don't think we need anything new from core. So, we just need to agree to throw our support behind a convention for loading modules. We should consider supporting Node.js and AMD.

For example, if we do what Node.js does, then we wrap the contents of a .js file in
```JS
(function (exports, require, module, __filename, __dirname) {
/* contents of JS file */
});
```
And then
```c
jerry_value_t the_function = jerry_eval (wrapped_file_contents);
jerry_call_function (the_function, ...);
```
and retrieve the exports.

AMD uses a different approach. For AMD we need not wrap the original JS in anything, but we need to provide the `define ()` method.
```JS
define( /* optional string with the module name */, [ "dep1", "dep2", "dep3" ],
  function( dep1, dep2, dep3 ) {
    /* Module body goes here. dep1, dep2, and dep3 are loaded asynchronously,
     * and the return value is the module. */
  } );
```

There's an effort to unify all of these into a single universal module definition: https://github.com/umdjs/umd

This is all a bit hairy, to say the least.